### DISCUSSION TITLE
Pocketbase in Azure

### DISCUSSION THREAD

**Discussion by EthanTuning** [pocketbase/pocketbase#2288] [Q&A] [answered]
Has anyone been able to successfully deploy Pocketbase to an Azure Container App, with a persistent storage solution?

I have been experimenting with hosting Pocketbase in an Azure Container App for quite some time now. I have a Docker file, Container Registry, and Azure File Share all setup and configured properly and am successful most of the way. I can run the Container App just fine, hit the Pocketbase 

But where I end up running into roadblocks when I try to mount an Azure File Share as a volume mount to the container. I do this because when you restart a Container App Revision, you lose all your data, since Containers are supposed to be stateless. When I try adding persistent storage, I end up getting the "SQLITE_BUSY (5)" error or the "out of memory (14)" error upon the Container App Revision firing up. 

I tried to dig into why this happens and come across lots of ideas, try most of them and nothing. It's extra weird to me because upon starting up the Container App Revision, the `data.db` file does pop up in the Azure File Share. So, it's as if everything works, yet the Container App fails to provision and then in turn won't deploy.

Has anyone come across issues like this? And a broader question has anyone successfully deployed Pocketbase to Azure Container Apps with a persistent storage solution?

**Comment by ganigeorgiev** [comment, marked as answer]
The `pb_data` must be mounted on a disk/volume - https://learn.microsoft.com/en-us/azure/virtual-machines/managed-disks-overview.

Unfortunately I have very little experience with the Azure ecosystem and I'm not sure if the above could be replaced with [Azure Files Share](https://learn.microsoft.com/en-us/azure/storage/files/storage-files-introduction).

Based on this stackoverflow answer outlining the differences between the 2 - https://stackoverflow.com/questions/41126309/azure-blob-vs-file-vs-disk-storage, it may not be possible to use Azure Files with PocketBase:

> Azure File Storage may be mounted as an SMB volume (so that all instances of your app can work with it). Note: This is not something readily supported with Web Apps currently - **you'd only be able to write to the file share via API, not via attached disk.**

**Reply by EthanTuning** [reply]
Thank you for snappy reply, I very much appreciate it and appreciate your hard work on this awesome tool!

Anyway, yeah I think the best thing for now is to just fire up a Ubuntu Server in Azure, and do everything a more standard way. Even though containerizing this would be much more handy.

It seems that it just won't work with how Azure Container Apps + Fire Share Volume Mount.

Just FYI too the Azure Container Apps resource is very different than the standard Azure VMs stuff. And, I really don't think they are far along with supporting the different thing you'd want to do with them.

**Reply by flashedk** [reply]
@EthanTuning what did you end up doing?

**Comment by Magnuti** [comment]
I had the exact same problems @EthanTuning! First with `out of memory (14)` and now with the `database is locked (5) (sqlite_busy)` error. I solved the first one by setting access mode on my file share to `ReadWrite` instead of `ReadOnly` lol. I haven't been able to solved the last error code yet, but I think using disk instead of Azure File Share will help. Did it work for you?

Here is my configuration #2745, I basically just use Azure Container Apps + Azure File Share.

**Reply by Magnuti** [reply]
Relevant threads where people are experiencing the same issue with Azure Files.
- https://github.com/microsoft/azure-container-apps/issues/18
- https://github.com/MicrosoftDocs/azure-docs/issues/47130

> The underlying storage is NFS/CIFS which is why SQLite isn't advise on a storage mount. The best workaround is to leverage either PostgreSQL or MySQL. https://github.com/MicrosoftDocs/azure-docs/issues/47130#issuecomment-1075237531

**Reply by flashedk** [reply]
@Magnuti what was your solution in the end?

**Reply by Magnuti** [reply]
I switched to Supabase.

**Comment by ccremer** [comment]
For anyone who is not shy to use Pocketbase as framework and can write Go code:

I was thinking to use [Event hooks](https://pocketbase.io/docs/event-hooks/) to upload data on Azure after a write operation, e.g. `OnModelAfter{Create,Update,Delete}`. Inside the hook, we'll connect to Azure Blob storage service and use [Azure Go SDK](https://github.com/Azure/azure-sdk-for-go/tree/main/sdk/storage/azblob) to upload the changed files.

@ganigeorgiev WDYT? would that be feasible?

Alternatively, integrate Azure Blob storage completely and make a feature for it (e.g. selectable in GUI as alternative to S3).

**Reply by ganigeorgiev** [reply]
It will be difficult to implement if you want to intercept the model hooks (_eventually this may get easier in the future with #2387_) and even if you manage to upload the files you'll have to later somehow retrieve them.

> Alternatively, integrate Azure Blob storage completely and make a feature for it (e.g. selectable in GUI as alternative to S3).

There are no plans to add other vendor specific storage apis as it will be hard to maintain (_even the current S3 abstraction is sometimes causing compatibility issues between vendors; see for example https://github.com/pocketbase/pocketbase/issues/2231_).

**Reply by ccremer** [reply]
Ok, my idea with the hook was to have a remote copy of the database, while the DB is still using the local files.
> you'll have to later somehow retrieve them.

Right, but we could download the remote copy before starting pocketbase, so that pocketbase has the local files again. Although, I have no idea if this approach is safe of file corruptions from transactions.

> There are no plans to add other vendor specific storage apis

Ack, understandable.

Thanks for the fast response :)

**Reply by ganigeorgiev** [reply]
The database must be always on the local filesystem. We don't support other storage mechanisms. The S3 configuration is only for the uploaded files and backups. If you want to replicate the db to S3 or other external servers you can check [litestream](https://litestream.io/), [marmot](https://github.com/maxpert/marmot), [litefs](https://github.com/superfly/litefs), etc.
